templates/terraform_setup.yml

```diff
@@ -17,44 +17,110 @@ parameters:
+    az network nsg create \
+      --resource-group ${RESOURCE_GROUP_NAME} \
+      --location       ${REGION} \
+      --name           ${NSG}
+    check_rc.sh $? "Create network security group (${NSG})" || exit 0
+
+    az network vnet create \
+      --resource-group         ${RESOURCE_GROUP_NAME} \
+      --location               ${REGION} \
+      --name                   ${VNET} \
+      --address-prefixes       ${VNET_ADDR_PREFIX} \
+      --subnet-name            ${SUBNET} \
+      --network-security-group ${NSG}
+    check_rc.sh $? "Create vnet (${VNET}) in ${RESOURCE_GROUP_NAME}" || exit 0
+
+    az network vnet subnet update \
+      --resource-group         ${RESOURCE_GROUP_NAME} \
+      --vnet-name              ${VNET} \
+      --name                   ${SUBNET} \
+      --service-endpoints      Microsoft.Storage
+    check_rc.sh $? "Add Microsoft.Storage to subnet ${SUBNET} in ${VNET}" || exit 0
+    echo "##[endgroup]"
+
+    # Prepare the cloud-init file.
+    cloud_init_file=${BUILD_SOURCESDIRECTORY}/templates/cloud_init/cloud_init.client.yml
+
+    base64_idrsa=$(base64 -w 0 ~/.ssh/id_rsa)
+    echo "##vso[task.setvariable variable=secret.SSH_ID_RSA_BASE64;issecret=true]${base64_idrsa}"
+    ssk_key_pub=$(cat ~/.ssh/id_rsa.pub)
+    echo "##vso[task.setvariable variable=secret.SSH_KEY_PUBLIC;issecret=true]${ssk_key_pub}"
+
+    new_node_count=$(( ${{ parameters.original_node_count }} + ${{ parameters.scale_up_node_count }} ))
+
+    sed -i -e "s/<id_rsa>/$base64_idrsa/" \
+      -e "s/<region>/$REGION/" \
+      -e "s/<avere_admin_pw>/$AVERE_ADMIN_PW/" \
+      -e "s/<sub_id>/$ARM_SUBSCRIPTION_ID/" \
+      -e "s/<arm_client_id>/$ARM_CLIENT_ID/" \
+      -e "s:<arm_client_scrt>:$ARM_CLIENT_SECRET:" \
+      -e "s/<arm_tenant_id>/$ARM_TENANT_ID/" \
+      -e "s/<resource_group_name>/$RESOURCE_GROUP_NAME/" \
+      -e "s/<vnet_name>/$VNET/" \
+      -e "s/<run_terraform_file>/$RUN_TERRAFORM_FILE/" \
+      -e "s,<ssh_key>,$(cat ~/.ssh/id_rsa.pub)," \
+      -e "s/<vm_admin_user>/azureuser/" \
+      -e "s/<controller_image>/${{ parameters.controller_image }}/" \
+      -e "s/<vfxt_image>/${{ parameters.image }}/" \
+      -e "s/<original_node_count>/${{ parameters.original_node_count }}/" \
+      -e "s/<new_node_count>/$new_node_count/" \
+      -e "s/<vdbench_config>/${{ parameters.vdbench_config }}/" \
+      -e "s,<vdbench_download_url>,${VDBENCH_URL}," \
+      ${cloud_init_file}
+    check_rc.sh $? "Preparing cloud-init file (sed)" || exit
+
+    # Deploy the client VMs.
+    az vm create -g ${RESOURCE_GROUP_NAME} \
+      -n ${VM_NAME}  \
+      --admin-username azureuser \
+      --vnet-name ${VNET} \
+      --subnet ${SUBNET} \
+      --image 'Canonical:0001-com-ubuntu-server-focal:20_04-lts:latest' \
+      --public-ip-address "" \
+      --size Standard_DS2_v2 \
+      --custom-data ${BUILD_SOURCESDIRECTORY}/templates/cloud_init/cloud_init.client.yml
```

Comment:
If I'm understanding the cloud-init file correctly, client setup is really client + cluster setup. If so, it should be clear that this isn't just a client setup, but involves the cluster as well. That way if there's a failure here, it's clear that the issue might be higher severity (cluster setup failure versus only a client issue, which is likely outside of Avere's control).

Also, if this is indeed client + cluster setup, then a success check is a must for cloud-init because a cluster setup failure should be very noticeable when reviewing a pipeline failure.

Please let me know if I'm misunderstanding the scope of this call.

Reply:
we talked about this - updated error msgs